### ISSUE TITLE
Instrumentation

### ISSUE THREAD

**Issue by jurre** [dependabot/dependabot-core#2504]
It would be great to have a way to instrument dependabot-core, to figure out performance bottlenecks etc.

There are a few options we could consider:

## Notifications

Introduce some sort of Notification pattern, so in or around a given method, we can do something like:
```ruby
def my_method
  Dependabot.instrument("bundler.my_method") do
    # .. the code
end
```

We could introduce some convenience methods so we can tag a given method as needing instrumentation.

Then, in the environment where we run this code, we could do something like:

```ruby
Dependabot::Instrumentation.subscribe("bundler.my_method") do |event|
  $statsd.timing("bundler.my_method", event.duration)
end
```

We could implement such an notification system ourselves, or include [ActiveSupport](https://api.rubyonrails.org/classes/ActiveSupport/Notifications.html)

## Allow setting a statsd/opentracing client

Alternatively we could inject a statsd or opentracing client into Dependabot, and ship with a noop implementation by default.

```ruby
Dependabot.statsd = Datadog::Statsd.new

def my_method
  time = Benchmark.measure do
    # ...
  end
  Dependabot.statsd.timing("bundler.my_method", time)
end
```

**Comment by jeffwidman** [comment]
👍 for adding [OpenTelemetry](https://opentelemetry.io/) support (opentracing was deprecated in favor of opentelemtry).

**Comment by abdulapopoola** [comment]
@JamieMagee , I believe you already added the framework for OTel so this might be light lift?

**Comment by JamieMagee** [comment]
Yes, the combination of migrating to `sentry-ruby`, which [supports OpenTelemetry][1], and adding OpenTelemetry configuration should make this easy to do.

I would think there are 2 steps:
- Configure Sentry to [propagate via OpenTelemetry][1]
- [Intstrument via OpenTelemetry][2]

[1]: https://docs.sentry.io/platforms/ruby/performance/instrumentation/opentelemetry/
[2]: https://opentelemetry.io/docs/languages/ruby/instrumentation/
